File: java/src/main/java/org/algorithm_visualizer/tracers/Array1DTracer.java
package org.algorithm_visualizer.tracers;

public class Array1DTracer extends Tracer {
    public static class Array1DOptions extends Options {
    }

    public Array1DTracer(String title, Array1DOptions options) {
        super(title, options);
    }

    public Array1DTracer(String title) {
        this(title, null);
    }

    public Array1DTracer(Array1DOptions options) {
        this(null, options);
    }

    public Array1DTracer() {
        this(null, null);
    }

    public Array1DTracer reset(Object... args) {
        Tracer.addTrace(key, "reset", args);
        return this;
    }

    public Array1DTracer set(Object... args) {
        Tracer.addTrace(key, "set", args);
        return this;
    }

    public Array1DTracer delay(Object... args) {
        Tracer.addTrace(key, "delay", args);
        return this;
    }

    public Array1DTracer patch(Object... args) {
        Tracer.addTrace(key, "patch", args);
        return this;
    }

    public Array1DTracer depatch(Object... args) {
        Tracer.addTrace(key, "depatch", args);
        return this;
    }

    public Array1DTracer select(Object... args) {
        Tracer.addTrace(key, "select", args);
        return this;
    }

    public Array1DTracer deselect(Object... args) {
        Tracer.addTrace(key, "deselect", args);
        return this;
    }

    public Array1DTracer chart(Object... args) {
        Tracer.addTrace(key, "chart", args);
        return (Array1DTracer) this;
    }
}

File: test/Array1DTracer.js
import test from 'ava';
import Tester from './helpers/Tester';
import { Array1DTracer, ChartTracer } from '..';

test('Array1DTracer', new Tester(execute => {
  let tracer;
  let key;
  let chartTracer;

  execute([
      tracer = new Array1DTracer('Test Tracer'),
      key = tracer.key,
      chartTracer = new ChartTracer(),
    ],
    { key, method: 'Array1DTracer', args: ['Test Tracer'] },
    { key: chartTracer.key, method: 'ChartTracer', args: [] },
  );

  execute([
      tracer.chart(chartTracer),
      tracer.chart(null),
    ],
    { key, method: 'chart', args: [chartTracer.key] },
    { key, method: 'chart', args: [null] },
  );

  execute([
      tracer.depatch(1),
    ],
    { key, method: 'depatch', args: [1] },
  );

  execute([
      tracer.deselect(2, 3),
    ],
    { key, method: 'deselect', args: [2, 3] },
  );

  execute([
      tracer.destroy(),
    ],
    { key, method: 'destroy', args: [] },
  );

  execute([
      tracer.patch(1, 'value'),
      tracer.patch(4, 999),
    ],
    { key, method: 'patch', args: [1, 'value'] },
    { key, method: 'patch', args: [4, 999] },
  );

  execute([
      tracer.reset(),
    ],
    { key, method: 'reset', args: [] },
  );

  execute([
      tracer.select(2, 3),
    ],
    { key, method: 'select', args: [2, 3] },
  );

  execute([
      tracer.set([1, 2, 3]),
      tracer.set(['a', 1, true]),
    ],
    { key, method: 'set', args: [[1, 2, 3]] },
    { key, method: 'set', args: [['a', 1, true]] },
  );
}).test);
